### PR TITLE
fix(eks): restricted public access breaks cluster functionality

### PR DESCRIPTION
**Pull request by iliapolo**
Currently, we attach the VPC to the `KubectlProvider` only when public access is not enabled. The idea was that if public access is enabled, the provider could always connect to the cluster via the internet. 

The problem is that public access can be restricted to specific CIDR's via the `onlyFrom` method. 

Solution is to switch up the logic and attach the VPC to the provider when private access is enabled. This would enable configuring `PUBLIC_AND_PRIVATE.onlyFrom(...)`. 

Also, using `PUBLIC.onlyFrom` is now unsupported because it will most likely break the provider since private access is disabled, and public access is restricted. 

Bottom line, these are the configurations that should work:

- Public (with or without private subnets)
- Private (with private subnets)
- Private and **unrestricted** public (with or without private subents) 
- Private and **restricted** public (with private subnets)

I also moved the `KubectlSecurityGroup` to be created only if needed.

Fixes https://github.com/aws/aws-cdk/issues/9866

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 1]
```diff
@@ -460,6 +460,11 @@ export class EndpointAccess {
    * @param cidr CIDR blocks.
    */
   public onlyFrom(...cidr: string[]) {
+    if (!this._config.privateAccess) {
+      // when private access is disabled, we can't restric public
+      // access since it will render the kubectl provider unusable.
+      throw new Error('Cannot restric public access to endpoint when private access is disabled. Use PUBLIC_AND_PRIVATE.onlyFrom() instead.');
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:466]: The terminology "restrict" is a bit vague. I would explicitly say something like "cannot specify cidr blocks for public access..."

**Review by eladb — APPROVED** [review]
